### DISCUSSION TITLE
How add new order types and vehicle type?

### DISCUSSION THREAD

**Discussion by Kicey** [openTCS/opentcs#16] [Q&A] [answered]
In the source code, there are only a few order type(defined in OrderConstants.java). They are not enough for my condition. How can I add new order type. Should I use Java SPI or guice module to do that, or the only way I can do that is adjust a lot of code? The same is adding new type of the vehicle. After adding them, how can I use that new types in the operation desk? Thansk.

**Comment by swltr** [comment, marked as answer]
The order types in `OrderConstants.java` are merely the default type strings available, to be shown in the dialog for creating transport orders in the Operations Desk application. You can set any type string you like/need when creating a transport order via that dialog or the API.

Vehicles are allowed to process transport orders of any type by default, indicated by a `*`. You can set a list of allowed order types for each individual vehicle at runtime, though - either manually in the Operations Desk application, or programmatically via `VehicleService.updateVehicleAllowedOrderTypes()`. The vehicle will only be assigned transport orders of types given in that list, then.

There is no explicit "vehicle type" concept in openTCS. You could say that every vehicle is individual (or its own type, if you will), so there is no need to add any types. You simply add another vehicle. (If this doesn't help, please explain what you're trying to achieve.)

**Reply by Kicey** [reply]
Understood, I think it's a good design that use string and self-setting order type to support to unknown order type needed by user. Thanks again!
